onnxruntime/core/providers/cpu/tensor/identity_op.h

```diff
@@ -43,7 +43,18 @@ class IdentityOp final : public OpKernel {
     }
 
     if (is_dropout) {
-      context->Output(1, std::vector<int64_t>());
+      Tensor* mask = context->Output(1, shape);
+      // a 'nullptr' returned would make it an unused optional output
+      if (mask != nullptr) {
+        // Opset 7 differs with Opset 10 in that the type of the 'mask'
+        // output is tied with the type of the input in Opset 7 whereas
+        // the type of 'mask' in Opset 10 is 'bool' always
+        // so we have a common solution
+        void* mask_data = mask->MutableDataRaw();
+        // In 'test'/'inference' mode, there are no input values dropped out
+        // so fill the buffer with 0/false
+        memset(mask_data, 0, mask->SizeInBytes());
```

Comment:
What if the tensor type is string?

Reply:
Input is constrained to a few numeric types so it shouldn't be.

Reply:
Then why " if (target != source)  " has such handling?

Reply:
I think Dropout doesn't support 'string' type (https://github.com/onnx/onnx/blob/master/docs/Operators.md#Dropout). 

For opset 8 onwards, the (optional) mask is always boolean. So not a problem here.

For opset 7, the (optional) mask is of the same type as the input tensor and the input types supported are float, double, float16. So string doesn't come into the picture here.

The question is - should I do something for float16 though ? (I suppose not)

Reply:
> Then why " if (target != source) " has such handling?

THat's because this kernel inherently supports 2 ops - Identity and Dropout. Identity supports string. Dropout does not. The new logic comes in the Dropout section.

Reply:
@hariharans29 : Doesn't float16 come for free anyway due to the implicit cast nodes? So, I don't think you need to "do something" special for it to work.

Reply:
You are right @fdwr :). No action needed.